OpenRA.Mods.Common/Traits/Armament.cs

```diff
@@ -141,15 +147,20 @@ public Armament(Actor self, ArmamentInfo info)
 
 		public virtual WDist MaxRange()
 		{
-			return new WDist(Util.ApplyPercentageModifiers(Weapon.Range.Length, rangeModifiers));
+			return new WDist(Util.ApplyPercentageModifiers(Weapon.Range.Length, rangeModifiers.Select(m => m.GetRangeModifier())));
 		}
 
 		protected override void Created(Actor self)
 		{
 			turret = self.TraitsImplementing<Turreted>().FirstOrDefault(t => t.Name == Info.Turret);
 			ammoPool = self.TraitsImplementing<AmmoPool>().FirstOrDefault(la => la.Info.Name == Info.AmmoPoolName);
 			coords = self.Trait<BodyOrientation>();
-			rangeModifiers = self.TraitsImplementing<IRangeModifier>().ToArray().Select(m => m.GetRangeModifier());
+			notifyBurstComplete = self.TraitsImplementing<INotifyBurstComplete>().ToArray();
+			reloadModifiers = self.TraitsImplementing<IReloadModifier>().ToArray();
+			rangeModifiers = self.TraitsImplementing<IRangeModifier>().ToArray();
+			damageModifiers = self.TraitsImplementing<IFirepowerModifier>().ToArray();
+			inaccuracyModifiers = self.TraitsImplementing<IInaccuracyModifier>().ToArray();
+			notifyAttacks = self.TraitsImplementing<INotifyAttack>().ToArray();
```

Comment:
I was assuming that it's the `.Select` part which needs to be updated every shot because `GetFoo()` returns something else if the modifier trait is disabled, for example.